include/wil/network.h

```diff
@@ -0,0 +1,1372 @@
+
+        return return_ipv6_address;
+    }
+
+    // non-member swap
+    inline void swap(::wil::network::socket_address& lhs, ::wil::network::socket_address& rhs) WI_NOEXCEPT
+    {
+        lhs.swap(rhs);
+    }
+
+    // class addr_info encapsulates the ADDRINFO-related structures returned from the socket functions
+    // getaddrinfo, GetAddrInfoW, GetAddrInfoWEx
+    // iterator semantics are supported to safely access these addresses
+    // template T supports pointers to the 3 address structures: ADDRINFOA*, ADDRINFOW*, ADDRINFOEXW*
+    template <typename T>
+    class addr_info_iterator_t
+    {
+    public:
+        // defining iterator_traits allows STL <algorithm> functions to be used with this iterator class.
+        // Notice this is a forward_iterator
+        // - does not support random-access (e.g. vector::iterator)
+        // - does not support bidirectional access (e.g. list::iterator)
+#if WIL_USE_STL || defined(WIL_DOXYGEN)
+        using iterator_category = ::std::forward_iterator_tag;
+#endif
+        using value_type = ::wil::network::socket_address;
+        using difference_type = size_t;
+        using distance_type = ptrdiff_t;
+        using pointer = const ::wil::network::socket_address*;
+        using reference = const ::wil::network::socket_address&;
+
+        constexpr addr_info_iterator_t() WI_NOEXCEPT = default;
+
+        explicit addr_info_iterator_t(const T* addrinfo) WI_NOEXCEPT :
+            // must const cast so we can re-use this pointer as we walk the list
+            m_current_ptr(const_cast<T*>(addrinfo))
+        {
+            if (m_current_ptr)
+            {
+                m_socket_address.reset(m_current_ptr->ai_addr, m_current_ptr->ai_addrlen);
+            }
+        }
+
+        const ::wil::network::socket_address& operator*() const WI_NOEXCEPT
+        {
+            return m_socket_address;
+        }
+
+        const ::wil::network::socket_address* operator->() const WI_NOEXCEPT
+        {
+            return &m_socket_address;
+        }
+
+        WI_NODISCARD bool operator==(const addr_info_iterator_t& rhs) const WI_NOEXCEPT
+        {
+            return m_current_ptr == rhs.m_current_ptr;
+        }
+
+        WI_NODISCARD bool operator!=(const addr_info_iterator_t& rhs) const WI_NOEXCEPT
+        {
+            return !(*this == rhs);
+        }
+
+        addr_info_iterator_t& operator++() WI_NOEXCEPT
+        {
+            WI_ASSERT(m_current_ptr);
+            if (m_current_ptr)
+            {
+                m_current_ptr = m_current_ptr->ai_next;
+
+                if (m_current_ptr)
+                {
+                    m_socket_address.reset(m_current_ptr->ai_addr, m_current_ptr->ai_addrlen);
+                }
+                else
+                {
+                    m_socket_address.reset();
+                }
+            }
+            else
+            {
+                m_socket_address.reset();
+            }

```

Comment:
```suggestion
            if (m_current_ptr)
            {
                m_current_ptr = m_current_ptr->ai_next;
            }

            if (m_current_ptr)
            {
                m_socket_address.reset(m_current_ptr->ai_addr, m_current_ptr->ai_addrlen);
            }
            else
            {
                m_socket_address.reset();
            }
```

A little simpler to read/follow and trivial for the optimizer to handle.

Alternatively you could probably drop the last `else` since `m_socket_address` should be in an empty state if `m_current_ptr` is null.